src/main/java/no/uib/cipr/matrix/sparse/ArpackGen.java

```diff
@@ -0,0 +1,261 @@
+        if (eigenvalues > matrix.numColumns() - 2 )
+            throw new IllegalArgumentException("NEV " + eigenvalues + " >= "
+                    + (matrix.numColumns())+" - 2");
+
+        int n = matrix.numRows();
+        intW nev = new intW(eigenvalues);
+
+        int ncv = Math.min(2 + eigenvalues + ncvModification, n); // Number of columns of the matrix V. NCV must satisfy the two inequalities 2 <= NCV-NEV and NCV <= N.
+
+        String bmat = "I";
+        String which = ritz.name();
+        doubleW tol = new doubleW(tolerance);
+        intW info = new intW(0);
+        int[] iparam = new int[11];
+        iparam[0] = 1;
+        iparam[2] = Math.max(MAX_ITERATIONS, n); // maximum number of Arnoldi update operations allowed (default = 300?)
+        iparam[4] = 1; // IPARAM(4) = NB: blocksize to be used in the recurrence. The code currently works only for NB = 1.
+        iparam[6] = 1; // No longer referenced. Implicit restarting is ALWAYS used.
+        iparam[7] = 1; // MODE On INPUT determines what type of eigenproblem is being solved. Must be 1,2,3,4; See under \Description of dnaupd  for the four modes available. (Mode 1:  A*x = lambda*x)
+        intW ido = new intW(0);
+
+        // used for initial residual (if info != 0)
+        // and eventually the output residual
+        double[] resid = new double[n];
+        // Lanczos basis vectors
+        double[] v = new double[n * ncv];
+        // Arnoldi reverse communication
+        double[] workd = new double[3 * n];
+        // private work array
+        double[] workl = new double[3*ncv * ncv + 6*ncv];
+        int[] ipntr = new int[11];
+
+        int i = 0;
+        while (true) {
+            i++;
+            arpack.dnaupd(ido, bmat, n, which, nev.val, tol, resid, ncv, v, n,
+                    iparam, ipntr, workd, workl, workl.length, info);
+            if (ido.val == 99){
+                if (info.val != 1) {
+                    break;
+                } else {
+                    // did not converge: Try lowering the accuracy:
+                    tol.val = tol.val * 10;
+                    System.out.println("New reduced tolerance: "+tol.val);
+                    return solve(eigenvalues, ritz, tol.val, ncvModification);
+                }
+            }
+            if (ido.val != -1 && ido.val != 1)
+                throw new IllegalStateException("ido = " + ido.val);
+            // could be refactored to handle the other types of mode
+            av(workd, ipntr[0] - 1, ipntr[1] - 1);
+        }
+
+        log.fine(i + " iterations for " + n);
+
+        if (info.val != 0 && info.val != 1) {
+            if (info.val == 3) {
+                //  3: No shifts could be applied during a cycle of the
+                //     Implicitly restarted Arnoldi iteration. One possibility
+                //     is to increase the size of NCV relative to NEV.
+                System.err.println("'No shifts could be applied during a cycle!'\n" +
```

Comment:
You're right, of course. I'll update this accordingly. Thanks for looking into it.